### PR TITLE
TFlu: Add Cortex-M generic target makefile

### PR DESCRIPTION
**Pull request by mansnils**
This is fixing issue: https://github.com/tensorflow/tensorflow/issues/43725

### CODE REVIEW

**Review by jenselofsson — DISMISSED** [review]

**Comment by gbaned** [comment]
@mansnils  Can you please resolve conflicts? Thanks!

**Comment by ml-0** [comment]
Hi @mansnils, I tried your PR. Yes, it goes into the right direction. I encountered two problems with it:

1.  The gcc executable was not found. I had to add the following lines to cortex_m_generic_makefile.inc
$(eval $(call add_third_party_download,$(GCC_EMBEDDED_URL),$(GCC_EMBEDDED_MD5),gcc_embedded,))
`export PATH := $(MAKEFILE_DIR)/downloads/gcc_embedded/bin/:$(PATH)
`

2. The CMSIS include directories were not added the INCLUDE. This happens if CMSIS_PATH is not populated. Question here: why does cmsis_nn.inc add the include dirs of Core, DSP, NN only of CMSIS_PATH is undefined?

**Comment by mansnils** [comment]
@ml-0 Thanks for your comments.

1) Will fix this.

2) My understanding is that INCLUDES are not allowed in cmsis.inc because Arduino can't handle it. See the RFC document for more details: https://docs.google.com/document/d/14GRxeVEgSKgKBKAijO7oxnI49nLoTYBFQmPok-rG0cw
And that's ok when using default CMSIS_PATH in tensorflow/lite/micro/tools/make/downloads/cmsis which gets patched. However when supplying a custom CMSIS_PATH when building, INCLUDES are needed. Assumption is that Arduino use the default path. (Actually jacking in a CMSIS_PATH have some other problems basically because it is unpatched but it is out of scope for this PR.)

File: tensorflow/lite/micro/tools/make/Makefile
Comment (advaitjain) [inline tensorflow/lite/micro/tools/make/Makefile:131]: based on https://github.com/tensorflow/tensorflow/issues/43865#issuecomment-705818053, can we have the toolchain be a separate flag instead of being part of `TAGS`?

Copying my suggestion here (and adding in some additional error checking):

```make
TOOLCHAIN :=gcc
ifeq ($(TOOLCHAIN), gcc))
  # default to gcc
  CXX_TOOL := g++
  CC_TOOL := gcc
  AR_TOOL := ar
else ifeq ($(TOOLCHAIN), armclang))
  CXX_TOOL := armclang++?
  CC_TOOL := ?
  AR_TOOL :=?
else
  $(error "Invalid TOOLCHAIN: "$(TOOLCHAIN))
endif
```

File: tensorflow/lite/micro/tools/make/ext_libs/cmsis_nn.inc
Comment (advaitjain) [inline tensorflow/lite/micro/tools/make/ext_libs/cmsis_nn.inc:15]: I gather that this is needed to allow @ml-0 to use CMSIS-NN that is not part of the downloads? Let's add an issue to the TODO to give more context on why this is needed.

Something like:
```
# TODO(#1234): Some use-cases need an upatched CMSIS-NN and these include paths
#  allow for that. In reality, patching is only needed for project generation (e.g. Arduino) so
#  it may be possible to always use an unpatched CMSIS-NN if there is a different approach
#  for project generation.
```

File: tensorflow/lite/micro/tools/make/ext_libs/cmsis_nn.inc
Comment (advaitjain) [inline tensorflow/lite/micro/tools/make/ext_libs/cmsis_nn.inc:132]: why are these additional headers needed?

File: tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc
Comment (advaitjain) [inline tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc:8]: With https://github.com/tensorflow/tensorflow/pull/43896 we are moving towards enforcing that the TARGET correspond to an actual directory. So we should go with `TARGET=cortex_m_generic` and have a separate command line option for the processor type (just like `CORTEX_M_CORE` in the previous version of the generic makefile).

File: tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc
Comment (advaitjain) [inline tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc:101]: I'm ok moving this to the top-level Makefile or leaving it here, whatever your preference. Would still like to use `TOOLCHAIN` instead of `TAGS`.

File: tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc
Comment (advaitjain) [inline tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc:123]: For consistency, let's use the `TARGET_TOOLCHAIN_PREFIX`, similar to https://github.com/tensorflow/tensorflow/blob/37da1f0ee19568ffddbb1b58d0f37bf8844427a6/tensorflow/lite/micro/tools/make/targets/bluepill_makefile.inc#L6

File: tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc
Comment (advaitjain) [inline tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc:141]: why `no-function-sections`? This is critical in reducing the binary size, at least for gcc.

File: tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc
Comment (advaitjain) [inline tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc:135]: not needed, it is part of the top-level Makefile now.

File: tensorflow/lite/micro/tools/make/ext_libs/cmsis_nn.inc
Comment (advaitjain) [inline tensorflow/lite/micro/tools/make/ext_libs/cmsis_nn.inc:12]: let's also add to the readme some of the various ways in which this particular target can be invoked, e.g. with an unpatched CMSIS outside of the TF Micro tree ...

File: tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc
Comment (advaitjain) [inline tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc:6]: I recognize that we haven't set a good example of documenting the Makefile usage. But this is likely a TARGET that will be used quite broadly. It would be great if you could move this documentation to the README that @ml-0 had started and provide more variations of the different ways to invoke this particular target.

**Review by advaitjain — CHANGES_REQUESTED** [review]
Thanks @mansnils. I think we're close. Let me know if some of my requested changes need more context or don't quite fit with your use-case.

> 2. My understanding is that INCLUDES are not allowed in cmsis.inc because Arduino can't handle it. See the RFC document for more details: https://docs.google.com/document/d/14GRxeVEgSKgKBKAijO7oxnI49nLoTYBFQmPok-rG0cw
>    And that's ok when using default CMSIS_PATH in tensorflow/lite/micro/tools/make/downloads/cmsis which gets patched. However when supplying a custom CMSIS_PATH when building, INCLUDES are needed. Assumption is that Arduino use the default path. (Actually jacking in a CMSIS_PATH have some other problems basically because it is unpatched but it is out of scope for this PR.)

If we need for this particular target to use an unpatched CMSIS, that is perfectly ok. Let's just add in context (inculding usage in the README and an issue for more details on what would be needed to support an unpatched CMSIS) so that we know why this is the case. The Arduino indeed needs the downloaded CMSIS to be patched.

**Comment by ml-0** [comment]
@mansnils, I found that the issue of missing CMSIS headers was due to an incomplete patching on my system (Win10/msys64). I filed a fix for it (#44001). After this is merged, I can use the new make target without adding the CMSIS include paths to INCLUDE.

File: tensorflow/lite/micro/tools/make/ext_libs/cmsis_nn.inc
Comment (ml-0) [reply tensorflow/lite/micro/tools/make/ext_libs/cmsis_nn.inc:15]: I was not trying to use an unpatched CMSIS but I accidentally used an incompletely patched CMSIS because of #44001. After #44001 is merged, I can use cmsis_nn.inc as @mansnils provided it. Anyway, the option of using an unpatched CMSIS might be useful.

Comment (advaitjain) [reply tensorflow/lite/micro/tools/make/ext_libs/cmsis_nn.inc:15]: ok, I'll let @mansnils decide what he prefers here -- I'm ok with allowing for unpatched CMSIS if the steps are documented in the README. Or removing that option for now and adding support on an as needed basis.

Comment (mansnils) [reply tensorflow/lite/micro/tools/make/ext_libs/cmsis_nn.inc:15]: OK, adding some info to the README.

File: tensorflow/lite/micro/tools/make/ext_libs/cmsis_nn.inc
Comment (mansnils) [reply tensorflow/lite/micro/tools/make/ext_libs/cmsis_nn.inc:132]: Without these the INCLUDE is needed as was there before. So those files are needed with TAGS=cmsis-nn.

File: tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc
Comment (mansnils) [reply tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc:101]: Done

File: tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc
Comment (mansnils) [reply tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc:123]: Done

File: tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc
Comment (mansnils) [reply tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc:141]: I am not sure. Perhaps a leftover. Removing it.

File: tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc
Comment (mansnils) [reply tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc:135]: Ok

File: tensorflow/lite/micro/tools/make/ext_libs/cmsis_nn.inc
Comment (mansnils) [reply tensorflow/lite/micro/tools/make/ext_libs/cmsis_nn.inc:12]: Done

File: tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc
Comment (mansnils) [reply tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc:6]: I will update the README.

File: tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc
Comment (advaitjain) [inline tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc:115]: use findstring instead?

```make
ifeq (cortex-m55, $(findstring cortex-m55, $(TARGET_ARCH)))
```

File: tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc
Comment (advaitjain) [inline tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc:138]: ```-DT_LITE_MCU_DEBUG_LOG``` should be part of all builds. Else we will have fprintf and abort calls in the library.

https://github.com/tensorflow/tensorflow/blob/b43c3c3a68e4b4e2c7fc4daa76c4f9485c7ed47a/tensorflow/lite/kernels/op_macros.h#L20-L46

File: tensorflow/lite/micro/cortex_m_generic/README.md
Comment (advaitjain) [inline tensorflow/lite/micro/cortex_m_generic/README.md:28]: Can you add backticks to fix up the markdown formatting?

\```
make -f tensorflow/lite/micro/tools/make/Makefile TOOLCHAIN=armclang TARGET=cortex_m_generic TARGET_ARCH=cortex-m55 microlite

make -f tensorflow/lite/micro/tools/make/Makefile TAGS=cmsis-nn TOOLCHAIN=armclang TARGET=cortex_m_generic TARGET_ARCH=cortex-m55 microlite

make -f tensorflow/lite/micro/tools/make/Makefile TAGS=cmsis-nn TOOLCHAIN=armclang TARGET=cortex_m_generic TARGET_ARCH=cortex-m55+nofp microlite

make -f tensorflow/lite/micro/tools/make/Makefile TOOLCHAIN=armclang TARGET=cortex_m_generic TARGET_ARCH=cortex-m7 microlite

make -f tensorflow/lite/micro/tools/make/Makefile TOOLCHAIN=armgcc TARGET=cortex_m_generic TARGET_ARCH=cortex-m7 microlite

make -f tensorflow/lite/micro/tools/make/Makefile TOOLCHAIN=armgcc TARGET=cortex_m_generic TARGET_ARCH=cortex-m4 microlite
\```

File: tensorflow/lite/micro/cortex_m_generic/README.md
Comment (advaitjain) [inline tensorflow/lite/micro/cortex_m_generic/README.md:31]: backticks here too.

File: tensorflow/lite/micro/tools/make/download_and_extract.sh
Comment (advaitjain) [inline tensorflow/lite/micro/tools/make/download_and_extract.sh:195]: I'm guessing https://github.com/tensorflow/tensorflow/pull/44014 will make this change not be necessary anymore? That should get merged today.

**Review by advaitjain — CHANGES_REQUESTED** [review]
I think once you have made the requested changes we can get this merged.

File: tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc
Comment (mansnils) [reply tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc:131]: I added -Wno-unused-private-field in the last commit, it is needed for:
tensorflow/lite/micro/simple_memory_allocator.h:100:18: error: private field 'error_reporter_' is not used [-Werror,-Wunused-private-field]
tensorflow/lite/micro/micro_allocator.cc:186:18: error: private field 'reporter_' is not used [-Werror,-Wunused-private-field]
Adding fomit-frame-pointer.

File: tensorflow/lite/micro/tools/ci_build/test_cortex_m_generic.sh
Comment (mansnils) [reply tensorflow/lite/micro/tools/ci_build/test_cortex_m_generic.sh:30]: Done

File: tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc
Comment (mansnils) [reply tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc:5]: Ok

File: tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc
Comment (mansnils) [reply tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc:76]: Yes

File: tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc
Comment (mansnils) [reply tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc:115]: This is trying to find all cortex-m55, e.g. cortex-m55+nofp, findstring won't work with pattern '%'.

File: tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc
Comment (mansnils) [reply tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc:138]: Ok

File: tensorflow/lite/micro/cortex_m_generic/README.md
Comment (mansnils) [reply tensorflow/lite/micro/cortex_m_generic/README.md:28]: Right

File: tensorflow/lite/micro/cortex_m_generic/README.md
Comment (mansnils) [reply tensorflow/lite/micro/cortex_m_generic/README.md:31]: Done

File: tensorflow/lite/micro/tools/make/download_and_extract.sh
Comment (mansnils) [reply tensorflow/lite/micro/tools/make/download_and_extract.sh:195]: Ok, removing it

File: tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc
Comment (mansnils) [reply tensorflow/lite/micro/tools/make/targets/cortex_m_generic_makefile.inc:108]: Ok

**Review by advaitjain — APPROVED** [review]

**Pull request merged**
